fzf.el

```diff
@@ -539,11 +558,17 @@ The returned lambda requires extra context information:
          (sh-cmd (concat fzf/executable " " args)))
     (with-current-buffer buf
       (setq default-directory (or directory "")))
-    (split-window-vertically window-height)
+
+    (unless fzf/position-bottom-of-frame
```

Comment:
I'm not sure I want to have that.  At least not without further customization control.  Without this, the fzf command shows at the bottom of my Emacs text-mode frame. With the code activated it splits the current window and shows up at the top of that window.  I end up with 3 windows.

Reply:
Hi,
first of all, thank you for your feedback! As you can see, I have no clue on how to develop Emacs packages. I want `fzf` to always stick to the bottom of my frame.

What I see on my end:
Example with `fzf/position-bottom-of-frame` set to `nil`, `fzf` invoked from `*scratch*`:
![image](https://github.com/user-attachments/assets/94418cea-942a-4797-8d3d-102819adb36f)
The buffer with `fzf` opens kind of in the middle of my frame, at the bottom of the buffer where it was invoked in.

Desired effect, with `fzf/position-bottom-of-frame` set to `t`, invoked from `*scratch*`:
![image](https://github.com/user-attachments/assets/1db9b467-5326-4c22-9904-a53d806d20c4)
`fzf` buffer sticks to the bottom of my frame, just above the minibuffer.

Reply:
Thanks for the screenshots.   I understand that you want the fzf display to always be at the bottom of the current frame regardless of the position of the current window. 

But the code as it is is creating an extra window when I use it.  If I start with a single window and execute fzf, I end up with 2 windows plus the fzf "window".   If I start with 2 windows I end up with 3 windows and the fzf "window".

I wonder if the code is correct.  I don't think you need a watcher for that.   

You probably need a helper function that can move the point to the bottom-most window (I would call it `fzf--windmove-bottom-most` or something similar.   That function could loop calling `windmove-down` until an exception is thrown.   It would be called early in the `fzf--start` when `fzf/position-bottom-of-frame` is non-nil to position the point where you want it.  It must be called before the window-height is computed.  Then the rest of the function would not change.  I did a small test and that seemed to work.

Reply:
Here's code that seems to work:

First, the function that moves point to the bottom-most window.  It loops until an exception (an error) occurs:

``` lisp
;; Internal helper function                                                                                                                                                                           
(defun fzf--move-to-bottom-window ()
  "Move point to the bottom-most window of the current frame"
  (condition-case nil
      (while 1
        (windmove-down 0))
    (error nil)))
```

Then all that needs to be done is to call that function when required, based on your new user-option,  and *before* the height of the window is computed.  One nice thing is this code is written in lisp and it's easy to inject execution of something even inside code that sets initial values, like the `let*` statement.   The `progn` form returns the result of `(window-height)` but if needed we first move the point to the the bottom window.  

Here's the new code for `fzf--start`. NOthing changed except the lines with the NEW comment:

``` lisp
;; Internal helper function                                                                                                                                                                           
(defun fzf--start (directory action &optional custom-args)
  "Launch `fzf/executable' in terminal, extract and act on selected item."
  (require 'term)

  ;; Clean up existing fzf, allowing multiple action types.                                                                                                                                           
  (fzf--close)

  (unless (executable-find fzf/executable)
    (user-error "Can't find fzf/executable '%s'. Is it in your OS PATH?"
                fzf/executable))

  ;; launch process in an inferior terminal mapped in current window                                                                                                                                  
  (window-configuration-to-register fzf--window-register)
  (advice-add 'term-handle-exit
              :after (fzf--after-term-handle-exit directory
                                                  action
                                                  fzf--target-validator
                                                  fzf--extractor-list))
  (let* ((term-exec-hook nil)
         (buf (get-buffer-create fzf/buffer-name))
         (min-height (min fzf/window-height
                          (/ (progn (when fzf/position-bottom-of-frame    ;; <<<<< NEW
                                     (fzf--move-to-bottom-window)).       ;; <<<<< NEW
                                   (window-height))
                             2)))
         (window-height (if fzf/position-bottom (- min-height) min-height))
         (args (or custom-args fzf/args))
         (sh-cmd (concat fzf/executable " " args)))
    (with-current-buffer buf
      (setq default-directory (or directory "")))
     (split-window-vertically window-height)
    (when fzf/position-bottom (other-window 1))
    (make-term (file-name-nondirectory fzf/executable)
               "sh" nil "-c" sh-cmd)
    (switch-to-buffer buf)

    ;; Disable minor modes that interfere with rendering while fzf is running                                                                                                                         
    ;; TODO: provide ability to modify the set of actions in the user-option                                                                                                                          
    ;;       to allow compatibility with more minor modes instead of using                                                                                                                            
    ;;       this hard coded set.                                                                                                                                                                     
    (and (fboundp 'turn-off-evil-mode) (turn-off-evil-mode))
    (when (and (fboundp 'linum-mode)
               (bound-and-true-p linum-mode))
      (linum-mode 0))
    (when (bound-and-true-p visual-line-mode)
      (visual-line-mode 0))
    (when (bound-and-true-p display-line-numbers-mode)
      (display-line-numbers-mode 0))
    ;; disable various settings known to cause artifacts, see #1 for more details                                                                                                                     
    (setq-local scroll-margin 0)
    (setq-local scroll-conservatively 0)
    (setq-local term-suppress-hard-newline t)
    (setq-local show-trailing-whitespace nil)
    (setq-local display-line-numbers nil)
    (setq-local truncate-lines t)
    (face-remap-add-relative 'mode-line '(:box nil))

    (and (fboundp 'term-char-mode) (term-char-mode))
    ;; Remember the used terminal exit handler to allow its later removal.                                                                                                                            
    (setq fzf--hook (fzf--after-term-handle-exit directory
                                                 action
                                                 fzf--target-validator
                                                 fzf--extractor-list)
          mode-line-format (format "   FZF  %s" (or directory "")))))
```

Reply:
That's really clean and understandable, thank you! Works great on my end. Code execution just before `window-height` is read is neat.

I'll make those changes now.

Reply:
I've found a downside of this new approach. When the bottom-most window is really small, splitting can just throw `size of new window too small`. On my end, this doesn't happen with the previous approach. Would you have some ideas on how to fix that?

Reply:
The code would have to look at the window-height before splitting.  The current system might be limited in this way anyway.  If there are x windows in 2x lines, it would be difficult to create another window by splitting one vertically.  The reason it's normally done in the current window is that it was assumed that the current window is tall enough to do something useful.  

One way might be to change the logic and find a window that has a minimum number of lines starting from where the current window is.  The function `fzf--move-to-bottom-window` could be renamed and could do a little bit more by checking the height of the new window and leave point into the window that is tall enough, trying to have it at the bottom.  It could first go at the bottom, check the height and if too small move up one and check again.  It would stop at where it stopped.  

Another possibility would be to create a root window at the bottom and use that without splitting any existing window and use the root window for the fzf.   It would always work but would violate the reason users normally use root windows (which is normally to create a window layout that is IDE-like).  That would require more modifications.

As I said at the beginning, I'm not sure I understand the original code and saw it creating more windows than required.  The `pop-to-buffer` way may be one way.  I just don't fully understand it at this point.  I saw from its docstring that it may pick another frame if there no good enough window.  If you use Emacs in terminal mode , only one frame is visible at a time, so that would not be very useful.   Using Emacs in terminal mode is very useful in several scenario, so you don't want to preclude this (but it could be checked dynamically).

Reply:
I see. I guess there wouldn't be anything bad in having an option that works only for the GUI mode? At this point it's purely your call though. I'll just push the version that you've described last time. 

I don't have the knowledge to get this to a really nice state, so feel free to close the PR. I assume you don't need this feature and working on it isn't something you'd do ;)

If you decide to close, people who'd like this option could just look into the two approaches we've tested and decide what works for them, it's not too much of a hassle to set up.

Again, thank you for all the feedback and exhaustive answers!

Reply:
I'd need to look into the mechanism using `pop-to-buffer` and understand it first.   The problem for me is the timing.  I'm very busy at the moment and will be for at least another month. I'll try to find time.  I think it's a useful option to have .  

In the mean time, can you tell me why you added a watcher to the variable?

Reply:
Just got an idea. I'll take a bit off time today. The idea is to identify the minimum allowed size (perhaps via a user-option that is customizable and cannot be smaller than something like 3) .

Then when looking for the window at the bottom, if it is too small, instead of splitting that window check if the use of a root window at the bottom is allowed (perhaps again via customization) and if there is no root window at the bottom create one and use that.  If there is already one root window, then just fail describing why.

I'll check if I can write something like that.

Reply:
OK, I have a preliminary version that finds the bottom-most window and check if it can split it.  If. it would be too small (current hard-coded limit of 5 lines) then it splits the frame in 2 parts creating a root window at the bottom and uses that.

The new code replaces
``` lisp
 (with-current-buffer buf
      (setq default-directory (or directory "")))
     (split-window-vertically window-height)
```

with:
``` lisp
    (with-current-buffer buf                                                                                                                                                                          
      (setq default-directory (or directory "")))                                                                                                                                                     
    (if (> (abs window-height) 5)                                                                                                                                                                     
        (progn                                                                                                                                                                                        
          (split-window-vertically window-height)                                                                                                                                                     
          (when fzf/position-bottom (other-window 1)))                                                                                                                                                
      (split-window (frame-root-window) nil 'below)                                                                                                                                                   
      (windmove-down 0))   
```

Reply:
It's possible that the resulting code has to be cleaned up a little, but everything seems to work.   Let me know if you find something wrong.

Reply:
See my fork at https://github.com/pierre-rouleau/fzf.el/blob/master/fzf.el; I updated it with the code from here for the moment.

Reply:
Hi, apologies for the late reply.
I've played around with your fork and couldn't break anything, so at least for my use case it works great. 

Thank your for the time you've spent to explain the approach and make this feature work!